app/services/submit_reference.rb

```diff
@@ -26,21 +27,28 @@ def save!
 
 private
 
-  def progress_application_if_enough_references_have_been_submitted
-    return unless there_are_now_enough_references_to_progress?
-
-    application_form.application_choices.each do |application_choice|
-      ApplicationStateChange.new(application_choice).references_complete!
-    end
-
-    SendApplicationsToProvider.new.call
-  end
-
   # Only progress the applications if the reference that is being submitted is
   # the 2nd referee, since there might be more than 2 references per form. We
   # don't want to send the references to the provider *again* when the 3rd or
   # 4th reference is submitted.
-  def there_are_now_enough_references_to_progress?
-    application_form.application_references.feedback_provided.count == ApplicationForm::MINIMUM_COMPLETE_REFERENCES
+  def enough_references_have_been_submitted?
+    (
+      application_form.application_references.feedback_provided + [@reference]
+    ).uniq.count == ApplicationForm::MINIMUM_COMPLETE_REFERENCES
+  end
+
+  def reference_feedback_provided!
+    @reference.update!(feedback_status: 'feedback_provided')
+  end
+
+  def progress_applications
+    ActiveRecord::Base.transaction do
+      reference_feedback_provided!
+      application_form.application_choices.each do |application_choice|
+        ApplicationStateChange.new(application_choice).references_complete!
+      end
+    end
+
+    SendApplicationsToProvider.new.call
```

Comment:
This service also writes to the database (it may be badly named) so there is still the potential for a partial failure here. Application choices could be moved to the `application_complete` state rather than all the way to `awaiting_provider_feedback`. I don't think this is necessarily a massive problem in this case because background processing would pick them up and progress them later. So I think this is OK, what do you think @duncanjbrown ?

If we are really concerned about having an atomic transaction then putting 10 second delay on the mail is the least messy alternative I can think of. Having separate services for DB and mail would also work but seems wrong to me.

Reply:
> I don't think this is necessarily a massive problem in this case because background processing would pick them up and progress them later.

My commit message could have been clearer in this regard, but yes - basically this. My thinking was also that the two services are loosely coupled, and that SubmitReference shouldn't concern itself with the state changes introduced by SendApplicationsToProvider. Any errors in the latter can be dealt with in isolation, with the knowledge that SubmitReference leaves application choices in the `application_complete` state.

Reply:
I agree, it's eventually consistent and that's a good thing